### ISSUE TITLE
CrossEntropyLoss masking

### ISSUE THREAD

**Issue by alrojo** [pytorch/pytorch#563]
The [CrossEntropyLoss](http://pytorch.org/docs/nn.html?highlight=crossentropyloss#torch.nn.CrossEntropyLoss) class and function uses inputs (unscaled probabilities), targets and class weights to calculate the loss.
The reason for using class weights is to help with imbalanced datasets.
However, with this setup you are not allowed to handle masking, which is a core issue in time-series (RNN, NLP) training with imbalanced sequence length.

I propose two alternative approaches that would solve this problem, while still allowing weighting class imbalances.

**1.** have the weights as a function argument at every execution. Such as [TensorFlow](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/seq2seq/python/ops/loss.py) is doing. Such that the weights is the same size as the target tensor, with a weighting on each of the samples. This would still allow the user to supply increased weight sizes at the indices of their imbalanced classes, while masking would be putting ones at wanted samples and zeros everywhere else.

**2.** A perhaps more elegant solution would be to have the CrossEntropyLoss exactly the same as tensorflows [cross entropy loss function](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/nn_ops.py#L1637), which seems to be the same as PyTorch's, but without averaging the loss of every sample. This would allow the user to average how they see fit and produce similar functions to the one in proposal (1).

**Comment by fmassa** [comment]
Hi,
Maybe I'm missing something here, but I have the impression that both of your points can be addressed using the [functional interface of `cross_entropy`](http://pytorch.org/docs/nn.html#torch.nn.functional.cross_entropy).
For **1**, we just need to pass the `weight` argument, which can be different for each input.
For **2**, we just pass `size_average=False` as the argument to `cross_entropy`.

**Comment by fmassa** [comment]
Ah, forget what I said, I misunderstood your statement, sorry.

**Comment by alrojo** [comment]
If you just changed the size_average from:

"size_average (bool, optional) – By default, the losses are averaged over observations for each minibatch. However, if the field sizeAverage is set to False, the losses are instead summed for each minibatch."

to

"size_average (bool, optional) – By default, the losses are averaged over observations for each minibatch. However, if the field sizeAverage is set to False, _the losses are a batch sized tensor with losses for each sample._"

You would be able to achieve (**2**) without breaking too many dependancies as it would be an optional argument.

**Comment by apaszke** [comment]
I have to say I quite like the 2nd approach. This would also fix #264. @soumith @colesbury wdyt?

**Comment by soumith** [comment]
i like the 2nd approach. Will take some reworking of our C / CUDA code though.

**Comment by jekbradbury** [comment]
Now that variable-length RNNs have been merged, it would be helpful to have a cross-entropy loss that allows masking, along the lines of Alex's second approach. But it isn't a showstopper because we can already get the same result with log_softmax followed by advanced indexing.

**Comment by jihunchoi** [comment]
Until this feature is supported in native C/CUDA, we can use [this code](https://gist.github.com/jihunchoi/f1434a77df9db1bb337417854b398df1), though sort of performance degradation is expected.

**Comment by ypxie** [comment]
What if if I need to apply weights to each different elements in each training sample? 
Why not have an element-wise loss function, with parameter to control whether to use mean or not?

**Comment by apaszke** [comment]
@ypxie the plan is to add element-wise losses.

**Comment by kvrd18** [comment]
Any updates on this? @apaszke

**Comment by kvrd18** [comment]
Can someone point me to the actual mathematical implementation of `NLLLoss` in PyTorch? I can try and manipulate it to compute element-wise loss. I had done it on Lua once.

**Comment by fmassa** [comment]
@kvrd18 here is the [C implementation](https://github.com/pytorch/pytorch/blob/master/torch/lib/THNN/generic/ClassNLLCriterion.c), and the CUDA implementation can be found in [here](https://github.com/pytorch/pytorch/blob/master/torch/lib/THCUNN/ClassNLLCriterion.cu) and [here](https://github.com/pytorch/pytorch/blob/master/torch/lib/THCUNN/generic/ClassNLLCriterion.cu).

**Comment by kvrd18** [comment]
@fmassa Thank you. I'll let you know once I get a working prototype.

**Comment by spro** [comment]
Would it make sense to add something like `masked_cross_entropy` from @jihunchoi to rnn utils?

**Comment by jekbradbury** [comment]
Yes, I think that would be a good idea.

**Comment by kvrd18** [comment]
@fmassa I've added a generic `WeightedClassNLLCriterion.c` in C and made the necessary definitions in `init.c` and `THNN.h`. Can you tell me how I can integrate it back all the way up to `nll_loss` like [`_functions.thnn.NLLLoss`](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L500)?

Also, @soumith, can you please add me to the PyTorch Slack?

**Comment by esube** [comment]
Is there any update on this other than the python level multiplication suggested by @jihunchoi ? This is an important feature for highly imbalanced data in multi-label case. For instance attribute recognition datasets are highly imbalanced from class to class (sometimes less than 1% positive examples per attribute) and are multi-label.

**Comment by kvrd18** [comment]
I've written C and CUDA functions for a generic `WeightedClassNLLCriterion`, which is an extension of `ClassNLLCriterion`. @vinayakvivek had to write new classes for calling this function from `auto.py` since it had to take an extra argument, `weight_maps`, when calling `forward` and `backward`. I'll issue a PR.

Regarding BCE, I think I'll have to write a new `InstanceWeightedLoss` class and get every criterion to inherit this class to pass 3 arguments for `forward` and `backward`. I believe this is going to be a needed feature if instance-wise weighted loss is going to be crucial.

@fmassa do you have any pointers for me regarding how to add this new common functionality for criteria?

**Comment by kvrd18** [comment]
The ideal scenario would be to avoid summation and provide a tensor of losses like how tensorflow does. I think this would require some major rewrites in the way losses are brought back to the front.

**Comment by karpathy** [comment]
I ran across this today too. Really what one wants is a tensor of the raw loss before any reduction (sum/mean as controlled by `size_average`). This is currently supported by TensorFlow's `tf.nn.sparse_softmax_cross_entropy_with_logits`, but not by PyTorch as far as I can tell. I tried implementing this manually with something along the lines of:

```
logp = F.log_softmax(net_out)
loss = -(per_pixel_weight * logp[:,labels]).mean() 
```

but this hangs indefinitely for some reason. Possible workarounds could be:

- ability to pass in a `weight` not per class, but per pixel, as determined automatically from shape of the `weight` input.
- perhaps using `size_average=None` to return the full tensor without the reduction. Feels a bit more gross.

**Comment by greydanus** [comment]
echoing @karpathy, a tensor of raw loss BEFORE any reduction would be great

**Comment by karpathy** [comment]
@greydanus  since it's all PyTorch in the implementation it's fine to do this manually without any performance hits. In fact, my performance even increased quite a bit because the CrossEntropyLoss does additional heavy lifting for ignoring of certain label values - a feature that most people won't use, but will pay in performance for. This almost seems like a poor choice for the library.

Anyway, assume `y_hat` is `(N,C)` class of raw scores (from a fully connected layer) and `y` are the true indices as `LongTensor`. We can do:

```python
logp = F.log_softmax(y_hat) # get (N,C) log probabilities
# prepare an (N,C) array of 1s at the locations of ground truth
ymask = logp.data.new(logp.size()).zero_() # (N,C) all zero
ymask.scatter_(1, y.view(-1,1), 1) # have to make y into shape (N,1) for scatter_ to be happy
ymask = Variable(ymask)
# pluck
logpy = (logp * ymask).sum(1) # this hurts in my heart
```

`logpy` becomes `(N,)` Tensor of the log probabilities of the correct classes. This works because during backpropagation the sum distributes the gradients equally to all channels, and then the gradient will get blocked everywhere where `ymask == 0` in each row, and will only flow through the elements of the correct classes. And if we wanted to weigh each example differently we would do

```python
negative_log_likelihood_loss = -(logpy * per_example_weights).mean()
```

**Comment by greydanus** [comment]
@karpathy thanks, that masking makes sense. I'm thinking about using this as a loss for a policy gradient algorithm. The idea would be to set `per_example_weights = discounted_ep_rewards`.

Sort of ironic you responded to this, as I was just reimplementing your **tf-agent** in PyTorch and trying to find an equivalent of `tf.nn.sparse_softmax_cross_entropy_with_logits`

**Comment by aosokin** [comment]
@karpathy , @greydanus Hi, I have a question related to masking the cross-entropy. Is there an easy way to compute log softmax when some labels are masked out?

For the moment, I had to implement in directly myself without using F.log_softmax. It seems that apart from performance losses this leads to some numerical instabilities, i.e., the training procedure starts to diverge from the path with F.log_softmax after 10-20 iterations. My current code:

```
def compute_argmax_masked(scores, mask):
    scores_for_max = scores.clone()
    scores_for_max.masked_fill_(mask == 0, float("-inf"))
    _, max_pos = torch.max(scores_for_max, 1, keepdim=True)
    return max_pos


def compute_max_masked(scores, mask):
    max_pos = compute_argmax_masked(scores, mask)
    return torch.gather(scores, 1, max_pos)


def compute_sum_masked(scores, mask):
    # not good: breaking the backprod chain (to get rid of infs):
    scores.data.masked_fill_(mask.data == 0, 0.0)
    # to get zero gradient in backprop, multiply by zero:
    scores = torch.mul(scores, mask.float())  # if one of the non-masked element is inf, this produces nans
    return scores.sum(1, keepdim=True)


def compute_log_softmax_masked(scores, mask):
    # find max values
    max_vals = compute_max_masked(scores, mask)
    # exponentiate
    scores_exp = torch.exp(scores - max_vals.expand_as(scores))
    # compute sum
    norm_vals = compute_sum_masked(scores_exp, mask)
    # take log
    norm_vals = torch.log(norm_vals) + max_vals
    # subtract sum
    scores = scores - norm_vals.expand_as(scores)
    return scores
```

**Comment by soumith** [comment]
I thought this through, and we will be introducing a:

```python
MyFavoriteLoss(..., size_average=..., sum=False)
```

The key being `sum=True` is the default summed loss (current behavior), and `sum=False` being unreduced / unsummed per-sample loss.

@zou3519 will be implementing this initially for a couple of losses (NLLLoss and MSELoss, and of course CrossEntropyLoss because it's LogSoftMax + NLLLoss), and it will be tracked in https://github.com/pytorch/pytorch/issues/264

Why not bootstrap `sizeAverage` here? because some losses have sizeAverage be per-sample normalization (based on total number of elements in the sample). An obvious example is MSELoss.

**Comment by fmassa** [comment]
As mentioned in https://github.com/pytorch/pytorch/issues/264#issuecomment-331120606, I'd rather not use the keyword `sum`, but something like `reduce` or `reduction`, because I think it might lead to confusion and is closer to TF naming.
What do you think?

**Comment by fmassa** [comment]
One more thing, I think that for some losses (like `MSELoss`, `L1Loss` and others) we should allow the target to require gradient, so that we can backprop through it.
This is not a big issue because those losses are easy to reimplement in one line, still would be a good thing to have I think.

**Comment by jihunchoi** [comment]
I think this feature is implemented in v0.3.0 in C/CUDA!
Though I haven't tested it yet. :P

**Comment by kvrd18** [comment]
Yes. The [`0.3`](http://pytorch.org/) release has an argument `reduce` in its loss functions, when set to `False` returns a tensor of losses.

Here is an example:

```python
import torch
import torch.nn as nn
from torch.autograd import Variable
inputs = Variable(torch.rand(5,2))
targets = Variable(torch.ones(5)).long()
print(nn.NLLLoss(reduce=False)(inputs, targets)) 

Variable containing:
-0.2527
-0.1675
-0.7303
-0.9321
-0.1608
[torch.FloatTensor of size 5]
```
